crates/bazed-input-mapper/src/key_combo.rs

```diff
@@ -0,0 +1,180 @@
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum KeySpec {
+    /// Key specified as a raw key code (<https://developer.mozilla.org/en-US/docs/Web/API/UI_Events/Keyboard_event_code_values>)
+    Raw(RawKey),
+    /// Key specified as a key attribute value (<https://www.w3.org/TR/uievents-key/>)
+    Str(Key),
+}
+
+impl std::fmt::Display for KeySpec {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Raw(s) => write!(f, "{}", s.0.replace("Digit", "").replace("Key", "")),
+            Self::Str(s) => write!(f, "{}", s.0),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Combo {
+    spec: KeySpec,
+    modifiers: Vec<Modifier>,
+}
+
+impl From<KeySpec> for Combo {
+    fn from(spec: KeySpec) -> Self {
+        Combo {
+            spec,
+            modifiers: Vec::new(),
+        }
+    }
+}
+
+impl Combo {
+    pub fn matches(&self, key_input: &KeyInput) -> bool {
+        let key_matches = match &self.spec {
+            KeySpec::Raw(x) => *x == key_input.code,
+            KeySpec::Str(x) => *x == key_input.key,
+        };
+        key_matches && self.modifiers == key_input.modifiers
+    }
+
+    pub fn from_keyinput_raw(input: KeyInput) -> Self {
+        Self {
+            modifiers: input.modifiers,
+            spec: KeySpec::Raw(input.code),
+        }
+    }
+    pub fn from_keyinput_str(input: KeyInput) -> Self {
+        Self {
+            modifiers: input.modifiers,
+            spec: KeySpec::Str(input.key),
+        }
+    }
+}
+
+impl std::ops::Add<Modifier> for RawKey {
+    type Output = Combo;
+
+    fn add(self, rhs: Modifier) -> Self::Output {
+        Combo {
+            spec: KeySpec::Raw(self),
+            modifiers: vec![rhs],
+        }
+    }
+}
+impl std::ops::Add<Modifier> for Combo {
+    type Output = Combo;
+
+    fn add(mut self, rhs: Modifier) -> Self::Output {
+        self.modifiers.push(rhs);
+        self
+    }
+}
+
+impl From<Key> for Combo {
+    fn from(key: Key) -> Self {
+        Self {
+            modifiers: Vec::new(),
+            spec: KeySpec::Str(key),
+        }
+    }
+}
+
+impl From<RawKey> for Combo {
+    fn from(key: RawKey) -> Self {
+        Self {
+            modifiers: Vec::new(),
+            spec: KeySpec::Raw(key),
+        }
+    }
+}
+
+impl std::fmt::Display for Combo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.modifiers.is_empty() {
+            write!(f, "{}", self.spec)
+        } else {
+            write!(f, "<{}-{}>", self.modifiers.iter().join("-"), self.spec)
+        }
+    }
+}
+
+impl FromStr for Combo {
+    type Err = KeyInputParseError;
+
+    // TODO related to normalization: Do we wanna parse `A` as `<S-A>`, `A` or `<S-a>`?
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Some(s) = s.strip_prefix('<').and_then(|s| s.strip_suffix('>')) {
```

Comment:
Does this explode when `<` without `>`?

Reply:
Explode how? It wouldn't parse that as a <thing>